File: sdk-api/src/main/java/dev/restate/sdk/Component.java
// Copyright (c) 2023 - Restate Software, Inc., Restate GmbH
//
// This file is part of the Restate Java SDK,
// which is released under the MIT license.
//
// You can find a copy of the license in file LICENSE in the root
// directory of this repository or package, or at
// https://github.com/restatedev/sdk-java/blob/main/LICENSE
package dev.restate.sdk;

import com.google.protobuf.ByteString;
import dev.restate.sdk.common.*;
import dev.restate.sdk.common.ComponentType;
import dev.restate.sdk.common.syscalls.*;
import java.util.*;
import java.util.concurrent.Executor;
import java.util.concurrent.Executors;
import java.util.function.BiFunction;
import java.util.stream.Collectors;

public final class Component implements BindableComponent<Component.Options> {
  private final ComponentDefinition<Component.Options> componentDefinition;
  private final Component.Options options;

  private Component(
      String fqsn, boolean isKeyed, HashMap<String, Handler<?, ?>> handlers, Options options) {
    this.componentDefinition =
        new ComponentDefinition<>(
            fqsn,
            isKeyed ? ComponentType.VIRTUAL_OBJECT : ComponentType.SERVICE,
            handlers.values().stream()
                .map(Handler::toHandlerDefinition)
                .collect(Collectors.toList()));
    this.options = options;
  }

  @Override
  public Options options() {
    return this.options;
  }

  @Override
  public List<ComponentDefinition<Component.Options>> definitions() {
    return List.of(this.componentDefinition);
  }

  public static ServiceBuilder service(String name) {
    return new ServiceBuilder(name);
  }

  public static VirtualObjectBuilder virtualObject(String name) {
    return new VirtualObjectBuilder(name);
  }

  public static class AbstractComponentBuilder {

    protected final String name;
    protected final HashMap<String, Handler<?, ?>> handlers;

    public AbstractComponentBuilder(String name) {
      this.name = name;
      this.handlers = new HashMap<>();
    }
  }

  public static class VirtualObjectBuilder extends AbstractComponentBuilder {

    VirtualObjectBuilder(String name) {
      super(name);
    }

    public <REQ, RES> VirtualObjectBuilder with(
        HandlerSignature<REQ, RES> sig, BiFunction<ObjectContext, REQ, RES> runner) {
      this.handlers.put(sig.getName(), new Handler<>(sig, runner));
      return this;
    }

    public Component build(Component.Options options) {
      return new Component(this.name, true, this.handlers, options);
    }
  }

  public static class ServiceBuilder extends AbstractComponentBuilder {

    ServiceBuilder(String name) {
      super(name);
    }

    public <REQ, RES> ServiceBuilder with(
        HandlerSignature<REQ, RES> sig, BiFunction<Context, REQ, RES> runner) {
      this.handlers.put(sig.getName(), new Handler<>(sig, runner));
      return this;
    }

    public Component build(Component.Options options) {
      return new Component(this.name, false, this.handlers, options);
    }
  }

  @SuppressWarnings("unchecked")
  public static class Handler<REQ, RES> implements InvocationHandler<Component.Options> {
    private final HandlerSignature<REQ, RES> handlerSignature;
    private final BiFunction<Context, REQ, RES> runner;

    public Handler(
        HandlerSignature<REQ, RES> handlerSignature,
        BiFunction<? extends Context, REQ, RES> runner) {
      this.handlerSignature = handlerSignature;
      this.runner = (BiFunction<Context, REQ, RES>) runner;
    }

    public HandlerSignature<REQ, RES> getHandlerSignature() {
      return handlerSignature;
    }

    public BiFunction<Context, REQ, RES> getRunner() {
      return runner;
    }

    public HandlerDefinition<Component.Options> toHandlerDefinition() {
      return new HandlerDefinition<>(
          this.handlerSignature.name,
          this.handlerSignature.requestSerde.schema(),
          this.handlerSignature.responseSerde.schema(),
          this);
    }

    @Override
    public void handle(
        Syscalls syscalls, Component.Options options, SyscallCallback<ByteString> callback) {
      // Wrap the executor for setting/unsetting the thread local
      Executor wrapped =
          runnable ->
              options.executor.execute(
                  () -> {
                    SYSCALLS_THREAD_LOCAL.set(syscalls);
                    try {
                      runnable.run();
                    } finally {
                      SYSCALLS_THREAD_LOCAL.remove();
                    }
                  });
      wrapped.execute(
          () -> {
            // Any context switching, if necessary, will be done by ResolvedEndpointHandler
            Context ctx = new ContextImpl(syscalls);

            // Parse input
            REQ req;
            try {
              req = this.handlerSignature.requestSerde.deserialize(syscalls.request().bodyBuffer());
            } catch (Error e) {
              throw e;
            } catch (Throwable e) {
              callback.onCancel(
                  new TerminalException(
                      TerminalException.BAD_REQUEST_CODE,
                      "Cannot deserialize input: " + e.getMessage()));
              return;
            }

            // Execute user code
            RES res;
            try {
              res = this.runner.apply(ctx, req);
            } catch (Error e) {
              throw e;
            } catch (Throwable e) {
              callback.onCancel(e);
              return;
            }

            // Serialize output
            ByteString serializedResult;
            try {
              serializedResult = this.handlerSignature.responseSerde.serializeToByteString(res);
            } catch (Error e) {
              throw e;
            } catch (Throwable e) {
              callback.onCancel(
                  new TerminalException(
                      TerminalException.INTERNAL_SERVER_ERROR_CODE,
                      "Cannot serialize output: " + e.getMessage()));
              return;
            }

            // Complete callback
            callback.onSuccess(serializedResult);
          });
    }
  }

  public static class HandlerSignature<REQ, RES> {

    private final String name;
    private final Serde<REQ> requestSerde;
    private final Serde<RES> responseSerde;

    HandlerSignature(String name, Serde<REQ> requestSerde, Serde<RES> responseSerde) {
      this.name = name;
      this.requestSerde = requestSerde;
      this.responseSerde = responseSerde;
    }

    public static <T, R> HandlerSignature<T, R> of(
        String method, Serde<T> requestSerde, Serde<R> responseSerde) {
      return new HandlerSignature<>(method, requestSerde, responseSerde);
    }

    public String getName() {
      return name;
    }

    public Serde<REQ> getRequestSerde() {
      return requestSerde;
    }

    public Serde<RES> getResponseSerde() {
      return responseSerde;
    }
  }

  public static class Options {
    public static final Options DEFAULT = new Options(Executors.newCachedThreadPool());

    private final Executor executor;

    /**
     * You can run on virtual threads by using the executor {@code
     * Executors.newVirtualThreadPerTaskExecutor()}.
     */
    public Options(Executor executor) {
      this.executor = executor;
    }
  }
}

File: sdk-api-kotlin/src/main/kotlin/dev/restate/sdk/kotlin/Component.kt
// Copyright (c) 2023 - Restate Software, Inc., Restate GmbH
//
// This file is part of the Restate Java SDK,
// which is released under the MIT license.
//
// You can find a copy of the license in file LICENSE in the root
// directory of this repository or package, or at
// https://github.com/restatedev/sdk-java/blob/main/LICENSE
package dev.restate.sdk.kotlin

import com.google.protobuf.ByteString
import dev.restate.sdk.common.BindableComponent
import dev.restate.sdk.common.ComponentType
import dev.restate.sdk.common.Serde
import dev.restate.sdk.common.TerminalException
import dev.restate.sdk.common.syscalls.*
import kotlin.coroutines.CoroutineContext
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import org.apache.logging.log4j.LogManager

class Component
private constructor(
    fqsn: String,
    isKeyed: Boolean,
    handlers: Map<String, Handler<*, *, *>>,
    private val options: Options
) : BindableComponent<Component.Options> {
  private val componentDefinition =
      ComponentDefinition(
          fqsn,
          if (isKeyed) ComponentType.VIRTUAL_OBJECT else ComponentType.SERVICE,
          handlers.values.map { obj: Handler<*, *, *> -> obj.toHandlerDefinition() })

  override fun options(): Options {
    return this.options
  }

  override fun definitions() = listOf(this.componentDefinition)

  companion object {
    fun service(
        name: String,
        options: Options = Options.DEFAULT,
        init: ServiceBuilder.() -> Unit
    ): Component {
      val builder = ServiceBuilder(name)
      builder.init()
      return builder.build(options)
    }

    fun virtualObject(
        name: String,
        options: Options = Options.DEFAULT,
        init: VirtualObjectBuilder.() -> Unit
    ): Component {
      val builder = VirtualObjectBuilder(name)
      builder.init()
      return builder.build(options)
    }
  }

  class VirtualObjectBuilder internal constructor(private val name: String) {
    private val handlers: MutableMap<String, Handler<*, *, ObjectContext>> = mutableMapOf()

    fun <REQ, RES> handler(
        sig: HandlerSignature<REQ, RES>,
        runner: suspend (ObjectContext, REQ) -> RES
    ): VirtualObjectBuilder {
      handlers[sig.name] = Handler(sig, runner)
      return this
    }

    inline fun <reified REQ, reified RES> handler(
        name: String,
        noinline runner: suspend (ObjectContext, REQ) -> RES
    ) = this.handler(HandlerSignature(name, KtSerdes.json(), KtSerdes.json()), runner)

    fun build(options: Options) = Component(this.name, true, this.handlers, options)
  }

  class ServiceBuilder internal constructor(private val name: String) {
    private val handlers: MutableMap<String, Handler<*, *, Context>> = mutableMapOf()

    fun <REQ, RES> handler(
        sig: HandlerSignature<REQ, RES>,
        runner: suspend (Context, REQ) -> RES
    ): ServiceBuilder {
      handlers[sig.name] = Handler(sig, runner)
      return this
    }

    inline fun <reified REQ, reified RES> handler(
        name: String,
        noinline runner: suspend (Context, REQ) -> RES
    ) = this.handler(HandlerSignature(name, KtSerdes.json(), KtSerdes.json()), runner)

    fun build(options: Options) = Component(this.name, false, this.handlers, options)
  }

  class Handler<REQ, RES, CTX : Context>(
      private val handlerSignature: HandlerSignature<REQ, RES>,
      private val runner: suspend (CTX, REQ) -> RES,
  ) : InvocationHandler<Options> {

    companion object {
      private val LOG = LogManager.getLogger()
    }

    fun toHandlerDefinition() =
        HandlerDefinition(
            handlerSignature.name,
            handlerSignature.requestSerde.schema(),
            handlerSignature.responseSerde.schema(),
            this)

    override fun handle(
        syscalls: Syscalls,
        options: Options,
        callback: SyscallCallback<ByteString>
    ) {
      val ctx: Context = ContextImpl(syscalls)

      val scope = CoroutineScope(options.coroutineContext)
      scope.launch {
        val serializedResult: ByteString

        try {
          // Parse input
          val req: REQ
          try {
            req = handlerSignature.requestSerde.deserialize(syscalls.request().bodyBuffer())
          } catch (e: Error) {
            throw e
          } catch (e: Throwable) {
            LOG.warn("Error when deserializing input", e)
            throw TerminalException(
                TerminalException.BAD_REQUEST_CODE, "Cannot deserialize input: " + e.message)
          }

          // Execute user code
          @Suppress("UNCHECKED_CAST") val res: RES = runner(ctx as CTX, req)

          // Serialize output
          try {
            serializedResult = handlerSignature.responseSerde.serializeToByteString(res)
          } catch (e: Error) {
            throw e
          } catch (e: Throwable) {
            LOG.warn("Error when serializing input", e)
            throw TerminalException(
                TerminalException.BAD_REQUEST_CODE, "Cannot serialize output: $e")
          }
        } catch (e: Throwable) {
          callback.onCancel(e)
          return@launch
        }

        // Complete callback
        callback.onSuccess(serializedResult)
      }
    }
  }

  class HandlerSignature<REQ, RES>(
      val name: String,
      val requestSerde: Serde<REQ>,
      val responseSerde: Serde<RES>
  )

  class Options(val coroutineContext: CoroutineContext) {
    companion object {
      val DEFAULT: Options = Options(Dispatchers.Default)
    }
  }
}
